qubes_config/widgets/gtk_widgets.py

```diff
@@ -0,0 +1,797 @@
+            )
+
+        if self.change_function:
+            self.change_function()
+
+        if self.style_changes:
+            self.entry_box.get_style_context().remove_class('combo-changed')
+            if self.is_changed():
+                self.entry_box.get_style_context().add_class('combo-changed')
+
+    def _apply_model(self):
+        assert isinstance(self.combo, Gtk.ComboBox)
+        list_store = Gtk.ListStore(int, str, GdkPixbuf.Pixbuf, str, str, str)
+
+        for entry_no, display_name in zip(itertools.count(),
+                                          sorted(self._entries)):
+            entry = self._entries[display_name]
+            list_store.append(
+                [
+                    entry_no,
+                    display_name,
+                    entry["icon"],
+                    entry["api_name"],
+                    '#f2f2f2' if entry['vm'] is None else None,  # background
+                    '#000000' if entry['vm'] is None else None,  # foreground
+                ])
+
+        self.combo.set_model(list_store)
+        self.combo.set_id_column(1)
+
+        icon_column = Gtk.CellRendererPixbuf()
+        self.combo.pack_start(icon_column, False)
+        self.combo.add_attribute(icon_column, "pixbuf", 2)
+        self.combo.set_entry_text_column(1)
+
+        entry_box = self.combo.get_child()
+
+        area = Gtk.CellAreaBox()
+        area.pack_start(icon_column, False, False, False)
+        area.add_attribute(icon_column, "pixbuf", 2)
+
+        completion = Gtk.EntryCompletion.new_with_area(area)
+        completion.set_inline_selection(True)
+        completion.set_inline_completion(True)
+        completion.set_popup_completion(True)
+        completion.set_popup_single_match(False)
+        completion.set_model(list_store)
+        completion.set_text_column(1)
+
+        entry_box.set_completion(completion)
+
+        # A Combo with an entry has a text column already
+        text_column: Gtk.CellRenderer = self.combo.get_cells()[0]
+        self.combo.reorder(text_column, 1)
+
+        # use list_store's 4th and 5th columns as source for background and
+        # foreground color
+        self.combo.add_attribute(text_column, 'background', 4)
+        self.combo.add_attribute(text_column, 'foreground', 5)
+
+        self.combo.connect("changed", self._combo_change)
+        self.entry_box.connect("changed", self._event_callback)
+
+    def _event_callback(self, *_args):
+        if self.change_function:
+            self.change_function()
+
+    def __str__(self):
+        return self.entry_box.get_text()
+
+    def get_selected(self) -> Optional[qubesadmin.vm.QubesVM]:
+        """
+        Get currently selected VM, if any
+        :return: QubesVM object
+        """
+        selected = self._get_valid_qube_name()
+
+        if selected in self._entries:
+            # special treatment for None:
+            if self._entries[selected]['api_name'] == "None":
```

Comment:
When I manually type `@tag:work` into source selection, it gets accepted, but displayed as `None`, and later in the policy it's also written as `None`. It should either support tags, or refuse to enter them. But not translate to "None". FWIW I tried it in policy for `qubes.Filecopy`.

PS not sure if this is specifically the place responsible, but looks relevant.

Reply:
....you can do that? apparently you can, ugh.